packages/sysutils/busybox/scripts/init

```diff
@@ -341,6 +341,65 @@
     $MOUNT_CMD "$MOUNT_TARGET" "$2" "$3" "$4"
   }
 
+  # mount the specified SYSTEM file and output arch from /etc/os-release
+  get_project_arch() {
+    mount_part "$1" "/sysroot" "ro,loop"
+
+    if [ -f /sysroot/etc/os-release ]; then
+      . /sysroot/etc/os-release
+      echo "${LIBREELEC_ARCH:-${OPENELEC_ARCH}}"
+    fi
+
+    umount /sysroot
+  }
+
+  # If the project/arch of current matches the upgrade, then it is considered compatible.
+  # Otherwise, mount the upgrade SYSTEM partition and, if canupdate.sh is available,
+  # call the script to determine if the current upgrade file can be applied on to the
+  # current system - 0 means it is compatible, non-zero that it is not compatible.
+  is_compatible() {
+    local result=1
+
+    if [ "${2}" = "${3}" ]; then
+      result=0
+    else
+      mount_part "$1" "/sysroot" "ro,loop"
+
+      if [ -f /sysroot/usr/share/bootloader/canupdate.sh ]; then
+        sh /sysroot/usr/share/bootloader/canupdate.sh "${2}" "${3}" && result=0
+      fi
+
+      umount /sysroot
+    fi
+
+    return ${result}
+  }
+
+  # determine if the new SYSTEM file is compatible with the current SYSTEM file
+  check_is_compatible() {
+    local old_project_arch="$(get_project_arch "${1}")"
+    local new_project_arch="$(get_project_arch "${2}")"
+
+    # If old or new project/arch isn't available then could be very old (pre-/etc/os-release) build - have to trust it
+    if [ -n "${old_project_arch}" -a "${new_project_arch}" ]; then
+      # If the old project/arch is not compatible with the new project/arch then abort...
+      if ! is_compatible "${2}" "${old_project_arch}" "${new_project_arch}"; then
+        echo ""
+        echo "ERROR: This update is not compatible with the current system - update aborted!"
```

Comment:
Is it possible to pass an error message and the filename from canupdate.sh back to the update script so the error message is more explicit? e.g.

`ERROR: LibreELEC-RPi3.aarch64-9.0.0.tar is not compatible with RPi2 hardware. Update cancelled.`